dbcon/joblist/jlf_tuplejoblist.cpp

```diff
@@ -3141,14 +3673,52 @@ void joinTablesInOrder(uint32_t largest, JobStepVector& joinSteps, TableInfoMap&
     thjs = dynamic_cast<TupleHashJoinStep*>(spjs.get());
     TupleBPS* tbps = dynamic_cast<TupleBPS*>(spjs.get());
 
-    if (*i == largest)
+    /* We have to specify for `TupleHasJoinStep` associated with key to be restored, to be on a large side,
+       to avoid situation with multiple small sides, when CS tries to merge join steps.
+
+       1. Consider the following join graph:
+           t2
+          /  \
+        t1 -- t3
+
+       {t3, t1} has a max weight, we choose that edge to remove from join graph.
+
+       2. Join graph after {t3, t1} edge was removed:
+           t2
+          /  \
+        t1    t3
+
+       a. It's legal for CS to schedule this joins as one step join as follow:
+       join(small sides {t1, t3}, large side {t2})
+
+       b. Instead of generating two steps join as follow:
+       t1_t2 = join(small sides {t1}, large side {t2})
+       join(small sides {t1_t2}, large side {t3)}
+
+       In case of `a` we unable to implement a join with a multiple keys.
```

Comment:
multiple keys that differs for t1_t2 and t2_t3 joins, right?

Reply:
I meant, that we could have a situation when we need a {t1, t2} to be on a small side as intemediate result of a previous join and t3 on a large side, in this case we can create a join for multiple keys, first keys represents an edge (t2, t3), and second keys represents an edge which we need to restore (t1, t3). But CS could apply an optimization and make a one large side {t2} and two small sides {t1, t3} in this case we cannot restore edge {t1. t3} because those 2 rowgroups (tables) on a small side, 2 different hash maps created for them.